### PR TITLE
Nodejs 4 WIP

### PR DESCRIPTION
**Pull request by heavyk**
do not merge yet. this is to track my progress

### CODE REVIEW

**Comment by piranna** [comment]
Ack :+1:

**Comment by heavyk** [comment]
I've found all kinds of stuff saying libsanitizer doesn't really work for musl (and other things too)

http://git.buildroot.net/buildroot/commit/?id=5f4d658d888b539de9a6247ae5b1a0999de5d4ec
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=56128

anyway, I've disabled it in the gcc configure statement.

cpio also needed to be updated with the new version of stdlibc++

**Comment by heavyk** [comment]
oh yeah, the first link there says `--disable-target-libsanitizer` and the second says `--disable-libsanitizer`

I added them both. seems to work

File: node_modules/nodeos-barebones/cpio.txt [hunk 1]
```diff
@@ -27,9 +27,9 @@ slink /lib/ld-musl-$CPU_FAMILY.so.1 libc.so                    0555 0 0
 # gcc
 
 file /lib/libgcc_s.so.1       $TOOLS/$TARGET/lib/libgcc_s.so.1       0444 0 0
-file /lib/libstdc++.so.6.0.17 $TOOLS/$TARGET/lib/libstdc++.so.6.0.17 0444 0 0
+file /lib/libstdc++.so.6.0.21 $TOOLS/$TARGET/lib/libstdc++.so.6.0.21 0444 0 0
 
-slink /lib/libstdc++.so.6 libstdc++.so.6.0.17 0444 0 0
+slink /lib/libstdc++.so.6 libstdc++.so.6.0.21 0444 0 0
```
Comment (piranna) [inline node_modules/nodeos-barebones/cpio.txt]: Thanks for remind me this, we need to find a better solution for versioned libraries :-) Maybe they could be injected the same way it's done with the `$TOOLS` and `$TARGET` variables...

**Comment by heavyk** [comment]
the other thing is too: I had some trouble with the way you download the node source. instead I just cloned the source from the git repo and then checked out `v4.1.1` -- compiles fine. now, I just need to figure out how to get this running :)

File: node_modules/nodeos-barebones/node_modules/nodeos-cross-toolchain/scripts/build [hunk 2]
```diff
@@ -112,6 +112,8 @@ if [[ ! -d $OBJ_DIR ]]; then
         --with-sysroot=$TOOLS/$TARGET         \
         --disable-nls                         \
         --disable-shared                      \
+        --disable-target-libsanitizer         \
+        --disable-libsanitizer                \
```
Comment (piranna) [inline node_modules/nodeos-barebones/node_modules/nodeos-cross-toolchain/scripts/build]: Maybe is http://lists.busybox.net/pipermail/buildroot/2014-May/095861.html related? We need to investigate the purposse of libsanitizier and to document if it's important or not.

**Comment by heavyk** [comment]
ok, I can't figure it out. what's the command to run it in qemu (or whatever)...

```
Setup is 16220 bytes (padded to 16384 bytes).
System is 11664 kB
CRC e1149d67
Kernel: arch/x86/boot/bzImage is ready  (#1)
Successfully compiled Linux
‘bzImage’ -> ‘/NodeOS/node_modules/nodeos-barebones/out/nocona/linux’
```

File: node_modules/nodeos-barebones/node_modules/nodeos-cross-toolchain/scripts/build (hunk 2 above)
Comment (heavyk) [inline node_modules/nodeos-barebones/node_modules/nodeos-cross-toolchain/scripts/build]: it's a tool similar to valgrind. it's an optional utility to ensure you're not going out of the memory area and for memory leaks. it's not used

Comment (heavyk) [inline node_modules/nodeos-barebones/node_modules/nodeos-cross-toolchain/scripts/build]: that link is the email which refers to this commit: http://git.buildroot.net/buildroot/commit/?id=5f4d658d888b539de9a6247ae5b1a0999de5d4ec

**Comment by piranna** [comment]
> what's the command to run it in qemu (or whatever)

`ǹpm start`. Each layer is independent, so if you only managed to compile nodeos-barebones, go into and exec it there, you'll get directly to a raw Node.js REPL.

File: node_modules/nodeos-barebones/node_modules/nodeos-cross-toolchain/scripts/build (hunk 2 above)
Comment (piranna) [inline node_modules/nodeos-barebones/node_modules/nodeos-cross-toolchain/scripts/build]: > it's a tool similar to valgrind

Thanks! :-D There's very little information about it besides was developed by Google...

**Comment by heavyk** [comment]
```
[ root@a67ed9c75d72:/NodeOS/node_modules {nodejs_4} ]$ apt-get install kvm
Reading package lists... Done
Building dependency tree
Reading state information... Done
Note, selecting 'qemu-kvm' instead of 'kvm'
The following NEW packages will be installed:
  qemu-kvm
0 upgraded, 1 newly installed, 0 to remove and 0 not upgraded.
Need to get 7,280 B of archives.
After this operation, 106 kB of additional disk space will be used.
Get:1 http://fi.archive.ubuntu.com/ubuntu/ vivid-updates/main qemu-kvm amd64 1:2.2+dfsg-5expubuntu9.4 [7,280 B]
Fetched 7,280 B in 0s (35.2 kB/s)
Selecting previously unselected package qemu-kvm.
(Reading database ... 42265 files and directories currently installed.)
Preparing to unpack .../qemu-kvm_1%3a2.2+dfsg-5expubuntu9.4_amd64.deb ...
Unpacking qemu-kvm (1:2.2+dfsg-5expubuntu9.4) ...
Processing triggers for man-db (2.7.0.2-5) ...
Setting up qemu-kvm (1:2.2+dfsg-5expubuntu9.4) ...
[ root@a67ed9c75d72:/NodeOS/node_modules {nodejs_4} ]$ npm start

> NodeOS@0.0.0 start /NodeOS
> scripts/start

Could not access KVM kernel module: No such file or directory
failed to initialize KVM: No such file or directory
```

**Comment by heavyk** [comment]
nevermind... I got it running by running the command

```
qemu-system-x86_64 -nographic --kernel bzImage -append console=ttyS0
```

File: node_modules/nodeos-barebones/cpio.txt (hunk 1 above)
Comment (heavyk) [inline node_modules/nodeos-barebones/cpio.txt]: should be ok as long as the gcc version stays the same. if you want to have multiple versions of gcc, just generate the file based on the symlink... `fs.readlink`

**Comment by piranna** [comment]
> Could not access KVM kernel module: No such file or directory failed to initialize KVM: No such file or directory

Definitely, it needs a `--help` flag...


> nevermind... I got it running

Congrats! :-D

**Comment by heavyk** [comment]
ok, it's just sitting at 100% cpu for the last 5 minutes. I assume that's not normal...

```
[    1.623105] Initializing XFRM netlink socket
[    1.625162] NET: Registered protocol family 10
[    1.630766] ip6_tables: (C) 2000-2006 Netfilter Core Team
[    1.632257] sit: IPv6 over IPv4 tunneling driver
[    1.634813] NET: Registered protocol family 17
[    1.635501] Key type dns_resolver registered
[    1.639476] registered taskstats version 1
[    1.642674]   Magic number: 15:791:597
[    1.643364] console [netcon0] enabled
[    1.643521] netconsole: network logging started
[    1.645556] ALSA device list:
[    1.645681]   No soundcards found.
[    1.670319] Freeing unused kernel memory: 7176K (ffffffff81aa6000 - ffffffff821a8000)
[    1.880391] tsc: Refined TSC clocksource calibration: 3399.978 MHz
[    1.880749] clocksource: tsc: mask: 0xffffffffffffffff max_cycles: 0x31023a3a8ac, max_idle_ns: 440795341378 ns
[    2.199404] input: ImExPS/2 Generic Explorer Mouse as /devices/platform/i8042/serio1/input/input3
[    2.883349] clocksource: Switched to clocksource tsc
```

**Comment by piranna** [comment]
> ok, it's just sitting at 100% cpu for the last 5 minutes. I assume that's not normal...

Uh-oh... definitely that's not good. Seems [the v8 version problem](https://github.com/nodejs/node-v0.x-archive/issues/9097) is still there, sh*t... :-( We'll need to give more pression on Node.js guys... :-/

**Comment by heavyk** [comment]
aha.... ok so it's a known issue. I'll have a look into where it's failing shortly. I'm gonna take a walk and eat dinner.

so close!!

**Comment by heavyk** [comment]
btw, this has been compiled with the musl git repo I linked in the other issue. I doubt that's the problem. if it is indeed a v8 issue, I hope I can get it.

how do I test running the node binary without running it in qemu?

**Comment by aulvi** [comment]
Great effort, keep pushing!

**Comment by piranna** [comment]
> I'm gonna take a walk and eat dinner.

Yeah, you've earned it ;-)


> btw, this has been compiled with the musl git repo I linked in the other issue. I doubt that's the problem

I don't think so, we have been using the latests stable releases until now without problems...


> if it is indeed a v8 issue, I hope I can get it

Seems so, there was a lot of people having problems with the v8 included on Node.js v0.11.15 and newer, but I though the io.js guys would have fixed it on Node.js 4.0.0 by using a bleeding-edge version...If the problem is still there then I don't know what would be happening :-/


> how do I test running the node binary without running it in qemu?

If you want to test *only* the Node.js binary on your Ubuntu machine, you can lauch it setting the `LD_LIBRARY_PATH` env var to the place where's located the compiled musl library from the toolchain, but as far as I remember this was working correctly...

**Comment by heavyk** [comment]
> we have been using the latests stable releases until now without problems...

I changed it when I was trying to figure out the RTLD_NEXT thing (cause of this commit https://github.com/bminor/musl/commit/6c5cad2aa56745302c1d42d2c8baf6424d29c0f3). I'll recompile again with stable to be sure.

I dunno how it even booted to be honest... I'm in the process of doing some conversions...

https://github.com/heavyk/node-century
https://github.com/heavyk/node-posix

**Comment by heavyk** [comment]
ok, here's the latest:

converted a few more packages that needed nan2
https://github.com/heavyk/node-src-ifaddrs
https://github.com/heavyk/node-src-errno
https://github.com/heavyk/node-src-sockios

all three of those belong to
https://github.com/heavyk/node-bin-ifconfig

the other main package that I had to modify to get the right dependencies was 
https://github.com/heavyk/nodeos-mount-rootfs

now I'm getting this strange error... I'll post the output. what happens is, I run npm install in `/NodeOS/node_modules/nodeos-usersfs/node_modules/bin-ifconfig` and when it gets to the spawning node-gyp part it spawns a million node processes (all defunct) and uses up all my ram. so, I cd into the directory manually and run npm install. you will see that it errors and cannot run the command.

if anyone has any ideas, I'd like help. I demonstrate the behaviour below:

```
[ root@a67ed9c75d72:/NodeOS/node_modules/nodeos-usersfs/node_modules/bin-ifconfig {nodejs_4 *} ]$ npm i
npm WARN package.json bin-ifconfig@0.0.0 No repository field.
/
> src-ifaddrs@0.0.0 install /NodeOS/node_modules/nodeos-usersfs/node_modules/bin-ifconfig/node_modules/src-ifaddrs
> node-gyp rebuild

^C
[ root@a67ed9c75d72:/NodeOS/node_modules/nodeos-usersfs/node_modules/bin-ifconfig {nodejs_4 *} ]$ cd node_modules/src-ifaddrs/
[ root@a67ed9c75d72:/NodeOS/node_modules/nodeos-usersfs/node_modules/bin-ifconfig/node_modules/src-ifaddrs {nodejs_4 *} ]$ npm i -dd
npm info it worked if it ends with ok
npm verb cli [ '/usr/local/bin/iojs', '/usr/local/bin/npm', 'i', '-dd' ]
npm info using npm@2.13.0
npm info using node@v2.4.0
npm verb readDependencies loading dependencies from /NodeOS/node_modules/nodeos-usersfs/node_modules/bin-ifconfig/node_modules/src-ifaddrs/package.json
npm WARN package.json src-ifaddrs@0.0.0 No repository field.
npm verb install where, deps [ '/NodeOS/node_modules/nodeos-usersfs/node_modules/bin-ifconfig/node_modules/src-ifaddrs',
npm verb install   [ 'nan' ] ]
npm verb install where, peers [ '/NodeOS/node_modules/nodeos-usersfs/node_modules/bin-ifconfig/node_modules/src-ifaddrs',
npm verb install   [] ]
npm verb installManyTop reading for lifecycle /NodeOS/node_modules/nodeos-usersfs/node_modules/bin-ifconfig/node_modules/src-ifaddrs/package.json
npm info preinstall src-ifaddrs@0.0.0
npm verb installManyTop reading scoped package data from /NodeOS/node_modules/nodeos-usersfs/node_modules/bin-ifconfig/node_modules/src-ifaddrs/node_modules/nan/package.json
npm verb readDependencies loading dependencies from /NodeOS/node_modules/nodeos-usersfs/node_modules/bin-ifconfig/node_modules/src-ifaddrs/package.json
npm verb targetResolver reading package data from /NodeOS/node_modules/nodeos-usersfs/node_modules/bin-ifconfig/node_modules/src-ifaddrs/node_modules/nan/package.json
npm verb already installed skipping nan@^2.0.9 /NodeOS/node_modules/nodeos-usersfs/node_modules/bin-ifconfig/node_modules/src-ifaddrs
npm info build /NodeOS/node_modules/nodeos-usersfs/node_modules/bin-ifconfig/node_modules/src-ifaddrs
npm info linkStuff src-ifaddrs@0.0.0
npm verb linkBins src-ifaddrs@0.0.0
npm verb linkMans src-ifaddrs@0.0.0
npm verb rebuildBundles src-ifaddrs@0.0.0
npm verb rebuildBundles [ 'nan' ]
npm info install src-ifaddrs@0.0.0
npm WARN cannot run in wd src-ifaddrs@0.0.0 node-gyp rebuild (wd=/NodeOS/node_modules/nodeos-usersfs/node_modules/bin-ifconfig/node_modules/src-ifaddrs)
npm info postinstall src-ifaddrs@0.0.0
npm info prepublish src-ifaddrs@0.0.0
npm verb validateInstall loading /NodeOS/node_modules/nodeos-usersfs/node_modules/bin-ifconfig/node_modules/src-ifaddrs/package.json for validation
npm verb exit [ 0, true ]
npm info ok
```

very strange... I'm off to bed.

@piranna - you said, it should be just as easy as changing just the version number :) --- lol

**Comment by piranna** [comment]
> I changed it when I was trying to figure out the RTLD_NEXT thing (cause of this commit bminor/musl@6c5cad2). I'll recompile again with stable to be sure.

Interesting, didn't know of that bug, good idea :-) Problem was that the symbols were not found, but disabling `libsanitizier` seems it did the trick, so it's ok.


> I dunno how it even booted to be honest... I'm in the process of doing some conversions...

If you only tested `nodeos-barebones`, there's no Node.js modules there at all, only Linux kernel booting directly to a Node.js REPL interpreter. For the others, nan 1.x seems to still be compatible with latests Node.js versions (1.9.0 it's just two months old), so there's not problem here. I asked for help to @netlovers to upgrade them to nan2 since he did the first implementation, but definitely you have done a huge work here :-D When you think they are ready I'll merge them.


> now I'm getting this strange error... I'll post the output. what happens is, I run npm install in /NodeOS/node_modules/nodeos-usersfs/node_modules/bin-ifconfig and when it gets to the spawning node-gyp part it spawns a million node processes (all defunct) and uses up all my ram

I think got something similar that was caused by node-gyp when they were set the modules as dependencies. Probably got fixed in npm3, but I'm not sure what was the reason of it.


> @piranna - you said, it should be just as easy as changing just the version number :) --- lol

Yeah, well... it's just only change the version number, but the importance is in the details... ;-)

**Comment by heavyk** [comment]
>  For the others, nan 1.x seems to still be compatible with latests Node.js versions (1.9.0 it's just two months old), so there's not problem here.

node-4.x is only compatible with nan-2.x

>  I asked for help to @netlovers to upgrade them to nan2 since he did the first implementation, but definitely you have done a huge work here :-D When you think they are ready I'll merge them.

I made 6 prs. one for each of the repos. I don't know if there are more. I haven't gotten that far (yet).

> I think got something similar that was caused by node-gyp when they were set the modules as dependencies. Probably got fixed in npm3, but I'm not sure what was the reason of it.

hmmm, ok, then today I will upgrade to the latest node/npm and try again (I'm using iojs-2.4.0 atm). I'll also check to see if node-gyp is a dep. I'm going to assume for the moment that it's a bug in npm though.

as it turns out, I hadn't compiled the userfs or the initramfs so with all of the correct(?) qemu options, I get this: (which is why I thought maybe the problem with the infinite loop was there).

```
[ root@a67ed9c75d72:/NodeOS {nodejs_4 *} ]$ qemu-system-x86_64 --kernel node_modules/nodeos-barebones/bzImage --initrd node_modules/nodeos-initramfs/initramfs.cpio.gz -hda node_modules/nodeos-usersfs/usersfs.img -vga std -append root=/dev/sda vga=extended -curses
qemu-system-x86_64: -hda node_modules/nodeos-usersfs/usersfs.img: could not open disk image node_modules/nodeos-usersfs/usersfs.img: Could not open 'node_modules/nodeos-usersfs/usersfs.img': No such file or directory
```

I can't get either of them to build (yet). I'll look again tonight. I have to finish up launching the website today.

cheers

**Comment by piranna** [comment]
> node-4.x is only compatible with nan-2.x

D'oh! Didn't know, I believed 1.9.0 was still compatible and it was only an API breakage, sorry.


> I made 6 prs. one for each of the repos. I don't know if there are more. I haven't gotten that far (yet).

I'll check and merge them, by far the only ones that I found that are being used and give problems have a fork on the [NodeOS](https://github.com/NodeOS) organization.


> it is a bug with npm. I just upgraded and it works properly

What version are you using now (so both could be able to use the same one)?


> however, it still complains about other things. I'll have a look tonight

Don't worry, take your time :-)


> as it turns out, I hadn't compiled the userfs or the initramfs so with all of the correct(?) qemu options, I get this: (which is why I thought maybe the problem with the infinite loop was there).
> 
> [ root@a67ed9c75d72:/NodeOS {nodejs_4 *} ]$ qemu-system-x86_64 --kernel node_modules/nodeos-barebones/bzImage --initrd node_modules/nodeos-initramfs/initramfs.cpio.gz -hda node_modules/nodeos-usersfs/usersfs.img -vga std -append root=/dev/sda vga=extended -curses
> qemu-system-x86_64: -hda node_modules/nodeos-usersfs/usersfs.img: could not open disk image node_modules/nodeos-usersfs/usersfs.img: Could not open 'node_modules/nodeos-usersfs/usersfs.img': No such file or directory
> 
> I can't get either of them to build (yet). I'll look again tonight

Although layers are independent they are stacked (based on Docker ones), so if you should try to generate (and test) in order: barebones, initramfs, rootfs and usersfs. initramfs will give you an "improved" environment with century compared to barebones (it's only mission is to mount the users filesystem), By the way, rootfs is now almost empty, and mostly only prepare a disk image to later burn it in an USB or CD-ROM.

**Comment by heavyk** [comment]
> What version are you using now (so both could be able to use the same one)?

the one that comes with 4.1.1...
```
$ node -v
v4.1.1
$ npm -v
2.14.4
```

> the only ones that I found that are being used and give problems have a fork on the NodeOS organization

ok, then there are a few to do... I'll do them later. I have a really badass script which does 90% of the conversion work for me.

> Although layers are independent they are stacked (based on Docker ones), so if you should try to generate (and test) in order: barebones, initramfs, rootfs and usersfs. initramfs will give you an "improved" environment with century compared to barebones (it's only mission is to mount the users filesystem), By the way, rootfs is now almost empty, and mostly only prepare a disk image to later burn it in an USB or CD-ROM.

aha, ok so I can begin debugging node then tonight to try and find that infinite loop. I'll try andding some printfs and set the commandline option `--log-all` to see where the bug happens. ok :+1: 

I need to get these working though because the initramfs is where core code will live (later, the rest will be downloaded into the disk image by node).

cheers, we're almost there :)

**Comment by heavyk** [comment]
> I believed 1.9.0 was still compatible and it was only an API breakage, sorry.

v8-4.3 introduced TypedArrays which is what all the buffer changes were (buffer uses typedarrays below the hood - and I think that was nan-1.{8,9}.x).. however, node-4.0 uses v8-4.5 and is completely incompatible because of the `MaybeLocal`s which were introduced in v8-4.4.
(don't quote me on any of this - it all moves so fast).

**Comment by piranna** [comment]
> What version are you using now (so both could be able to use the same one)?
> 
> the one that comes with 4.1.1...

Ok, I was having Node.js 0.12.7 and npm 2.13.2, I've just updated the last one... just to find that npm3 has been released and now I have npm 3.3.3 :-P That are good news :-)


> I have a really badass script which does 90% of the conversion work for me

Is it a personal one? Is it published somewhere?


> I need to get these working though because the initramfs is where core code will live

A disk-less system? Yeah, seems a good idea for your use case :-) Take in account that if so, apps will be running as root instead of unpriviledget chroot by default, but I don't think it would be difficult to fix it.


> later, the rest will be downloaded into the disk image by node

By using npm? Downloading it like PXE? If you use only the initramfs and instances will be dedicated, you could all everything to the initram image and boot directly from there without needing to download anything... Also npm itself crash when running it inside NodeOS, don't know if it's an out of memory problem or something...


> cheers, we're almost there :)

Yeah, you are doing a great job! :-D

**Comment by heavyk** [comment]
quick update: no real progress last night. this weekend I'm really busy doing stuff for the website. however, it is my expectation to make progress again on monday.

have a good weekend!

**Comment by piranna** [comment]
Don't worry, work is first :-)

I'm working on integrating [ExclFS](https://github.com/NodeOS/ExclFS) for the `/dev` folder, I hope have it today or this weekend. When I'll finish this I'll review your pull-request.

Good weekend to everybody too! Enjoy! :-)

**Comment by piranna** [comment]
Just one thing: DON'T USE NPM@3!!! When yesterday I upgraded it, since it try to use a plain hierarchy it removed the nodeos-cross-tools module. npm@3 has some obvious advantages for us, but seems porting to it will need some thought work and testing... :-/

**Comment by piranna** [comment]
@heavyk, on Ubuntu 15.10 it's included GCC 5.2 that has problems to compile GCC 4.7.3, so I've managed to update it by cherry-picking some of your commits. Please do a rebase of your branch over master :-)

**Comment by lite20** [comment]
Does this branch need to remain? Seeing that we now have nodejs4?

**Comment by piranna** [comment]
Yes, we can close it, in fact we are using @heavyk patches :-)

**Pull request closed without merging**